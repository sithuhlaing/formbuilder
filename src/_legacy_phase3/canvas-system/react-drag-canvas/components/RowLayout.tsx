/**
 * RowLayout Component
 * Renders a row layout container with drag-drop support
 * This is specifically for horizontal row layouts (different from HorizontalLayout)
 */

import React, { useRef } from 'react';
import { useDrop, useDrag } from 'react-dnd';
import { SmartDropZone } from './SmartDropZone';
import type { CanvasItem, DragItem } from '../types';

interface RowLayoutProps {
  item: CanvasItem;
  index: number;
  renderItem: (item: CanvasItem, context: any) => React.ReactNode;
  onItemDelete: (itemId: string) => void;
  onItemMove?: (dragIndex: number, hoverIndex: number) => void;
  onLayoutCreate?: (componentType: string, targetId: string, side: 'left' | 'right') => void;
  onAddToLayout?: (itemType: string, layoutId: string) => void;
  selectedItemId?: string;
  cssPrefix?: string;
  config?: any;
}

export const RowLayout: React.FC<RowLayoutProps> = ({
  item,
  index,
  renderItem,
  onItemDelete,
  onItemMove,
  onLayoutCreate,
  onAddToLayout,
  selectedItemId,
  cssPrefix = 'canvas',
  config
}) => {
  const children = item.children || [];
  const ref = useRef<HTMLDivElement>(null);

  // Drag functionality for the entire row layout - constrained to vertical movement only
  const [{ isDragging }, drag] = useDrag({
    type: 'existing-item',
    item: {
      type: 'existing-item',
      id: item.id,
      sourceId: item.id,
      item: item,
      dragType: 'row-layout', // This identifies it as a row layout for constraint application
      index: index
    },
    collect: (monitor) => ({
      isDragging: monitor.isDragging()
    })
  });

  // Drop functionality for the row layout container
  const [{ isOver }, drop] = useDrop({
    accept: ['new-item', 'existing-item'],
    drop: (dragItem: DragItem, monitor) => {
      if (monitor.didDrop()) return; // Prevent duplicate drops
      
      // Handle drops within row layout - add to existing layout instead of creating new one
      const componentType = dragItem.itemType || (dragItem.item && dragItem.item.type);
      if (componentType && onAddToLayout) {
        console.log('RowLayout: Adding component to existing layout:', componentType, item.id);
        onAddToLayout(componentType, item.id);
      }
    },
    collect: (monitor) => ({
      isOver: monitor.isOver({ shallow: true })
    })
  });

  // Combine drag and drop refs with the component ref
  drag(drop(ref));

  return (
    <div 
      ref={ref}
      className={`${cssPrefix}__row-layout row-layout ${isOver ? 'is-drop-target' : ''} ${isDragging ? 'is-dragging' : ''}`}
      data-testid="row-layout"
      data-component-id={item.id}
      data-component-type={item.type}
      data-layout-index={index}
      style={{
        display: 'flex',
        flexDirection: 'row',
        gap: '12px',
        padding: '8px',
        border: '1px solid #e0e0e0',
        borderRadius: '4px',
        backgroundColor: isOver ? '#f0f9ff' : '#fafafa',
        minHeight: '60px',
        alignItems: 'stretch',
        position: 'relative',
        opacity: isDragging ? 0.7 : 1,
        cursor: 'move'
      }}
    >
      {/* Layout header with drag handle */}
      <div className={`${cssPrefix}__layout-header`} style={{
        position: 'absolute',
        top: '-8px',
        left: '8px',
        backgroundColor: '#fff',
        padding: '2px 8px',
        fontSize: '12px',
        color: '#666',
        border: '1px solid #e0e0e0',
        borderRadius: '4px',
        display: 'flex',
        alignItems: 'center',
        gap: '8px'
      }}>
        <span className={`${cssPrefix}__drag-handle`} style={{
          cursor: 'move',
          color: '#9ca3af',
          fontSize: '14px',
          lineHeight: 1,
          userSelect: 'none'
        }}>
          ⋮⋮
        </span>
        <span className={`${cssPrefix}__layout-label`}>
          Row Layout ({children.length} items)
        </span>
        <button
          className={`${cssPrefix}__delete-btn`}
          onClick={(e) => {
            e.stopPropagation();
            onItemDelete?.(item.id);
          }}
          aria-label="Delete layout"
          style={{
            background: 'none',
            border: 'none',
            color: '#ef4444',
            cursor: 'pointer',
            fontSize: '14px',
            padding: '0',
            width: '16px',
            height: '16px',
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center'
          }}
        >
          ×
        </button>
      </div>

      {/* Layout content - horizontal flex container */}
      <div className={`${cssPrefix}__layout-content`} style={{
        display: 'flex',
        flexDirection: 'row',
        gap: '12px',
        flex: 1,
        alignItems: 'stretch',
        minHeight: '50px'
      }}>
        {children.map((child, childIndex) => (
          <div
            key={child.id}
            data-testid={`row-item-${childIndex}`}
            className={`${cssPrefix}__layout-item`}
            style={{
              flex: '1',
              minWidth: '120px',
              display: 'flex',
              flexDirection: 'column'
            }}
          >
            <SmartDropZone
              item={child}
              index={childIndex}
              renderItem={renderItem}
              onItemDelete={onItemDelete}
              onItemMove={onItemMove || (() => {})}
              onLayoutCreate={onLayoutCreate || (() => {})}
              selectedItemId={selectedItemId}
              config={config}
            />
          </div>
        ))}
        
        {/* Drop zone for adding new items to the end of the row */}
        {children.length < 4 && (
          <div 
            className={`${cssPrefix}__add-to-row`}
            style={{
              flex: '0 0 120px',
              minHeight: '50px',
              border: '2px dashed #d1d5db',
              borderRadius: '4px',
              display: 'flex',
              alignItems: 'center',
              justifyContent: 'center',
              color: '#9ca3af',
              fontSize: '12px',
              backgroundColor: isOver ? '#e0f2fe' : 'transparent',
              pointerEvents: 'none' // Prevent interference with container drop
            }}
          >
            Drop here to add
          </div>
        )}
      </div>
    </div>
  );
};
