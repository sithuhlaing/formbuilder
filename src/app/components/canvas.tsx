"use client";
import React, { useState } from "react";
import TextInputRenderer from "./renderers/TextInputRenderer";
import SelectRenderer from "./renderers/SelectRenderer";
import DefaultRenderer from "./renderers/DefaultRenderer";
import EmailInputRenderer from "./renderers/EmailInputRenderer";
import PasswordInputRenderer from "./renderers/PasswordInputRenderer";
import NumberInputRenderer from "./renderers/NumberInputRenderer";
import TextAreaRenderer from "./renderers/TextAreaRenderer";
import RichTextRenderer from "./renderers/RichTextRenderer";
import MultiSelectRenderer from "./renderers/MultiSelectRenderer";
import CheckboxRenderer from "./renderers/CheckboxRenderer";
import RadioGroupRenderer from "./renderers/RadioGroupRenderer";
import DatePickerRenderer from "./renderers/DatePickerRenderer";
import FileUploadRenderer from "./renderers/FileUploadRenderer";
import SignatureRenderer from "./renderers/SignatureRenderer";
import HorizontalLayoutRenderer from "./renderers/HorizontalLayoutRenderer";
import VerticalLayoutRenderer from "./renderers/VerticalLayoutRenderer";
import SectionDividerRenderer from "./renderers/SectionDividerRenderer";
import ButtonRenderer from "./renderers/ButtonRenderer";
import HeadingRenderer from "./renderers/HeadingRenderer";
import CardRenderer from "./renderers/CardRenderer";

const componentRenderers: { [key: string]: React.ComponentType<any> } = {
  text_input: TextInputRenderer,
  email_input: EmailInputRenderer,
  password_input: PasswordInputRenderer,
  number_input: NumberInputRenderer,
  textarea: TextAreaRenderer,
  rich_text: RichTextRenderer,
  select: SelectRenderer,
  multi_select: MultiSelectRenderer,
  checkbox: CheckboxRenderer,
  radio_group: RadioGroupRenderer,
  date_picker: DatePickerRenderer,
  file_upload: FileUploadRenderer,
  signature: SignatureRenderer,
  horizontal_layout: HorizontalLayoutRenderer,
  vertical_layout: VerticalLayoutRenderer,
  section_divider: SectionDividerRenderer,
  button: ButtonRenderer,
  heading: HeadingRenderer,
  card: CardRenderer,
};

export default function Canvas() {
  const [droppedComponents, setDroppedComponents] = useState<any[]>([]);

  const onDragOver = (e: React.DragEvent) => {
    e.preventDefault();
  };

  const onDrop = (e: React.DragEvent) => {
    e.preventDefault();
    const componentData = JSON.parse(e.dataTransfer.getData("application/json"));
    console.log("Dropped component:", componentData);
    setDroppedComponents([...droppedComponents, componentData]);
  };

  const renderComponent = (component: any): React.ReactNode => {
    console.log("Rendering component:", component.type, "Available renderers:", Object.keys(componentRenderers));
    const Renderer = componentRenderers[component.type] || DefaultRenderer;

    // Handle layout components with nested children
    if (component.type === 'horizontal_layout' && component.columns) {
      return (
        <Renderer component={component}>
          {component.columns.map((column: any, colIndex: number) => (
            <div key={colIndex} className="flex-1">
              {column.fields?.map((field: any, fieldIndex: number) => (
                <div key={fieldIndex}>
                  {renderComponent(field)}
                </div>
              ))}
            </div>
          ))}
        </Renderer>
      );
    }

    if (component.type === 'vertical_layout' && component.fields) {
      return (
        <Renderer component={component}>
          {component.fields.map((field: any, fieldIndex: number) => (
            <div key={fieldIndex}>
              {renderComponent(field)}
            </div>
          ))}
        </Renderer>
      );
    }

    if (component.type === 'card' && component.children) {
      return (
        <Renderer component={component}>
          {component.children.map((child: any, childIndex: number) => (
            <div key={childIndex}>
              {renderComponent(child)}
            </div>
          ))}
        </Renderer>
      );
    }

    // For regular components (input fields, etc.)
    return <Renderer component={component} />;
  };

  return (
    <div
      onDragOver={onDragOver}
      onDrop={onDrop}
      className="flex-1 bg-gray-100 dark:bg-gray-800 p-8"
    >
      <div className="bg-white dark:bg-gray-900 rounded-lg shadow-md p-6 min-h-full border-2 border-dashed border-gray-300 dark:border-gray-700">
        {droppedComponents.length > 0 ? (
          droppedComponents.map((component, index) => (
            <div key={index}>{renderComponent(component)}</div>
          ))
        ) : (
          <div className="flex items-center justify-center h-full">
            <div className="text-center text-gray-500 dark:text-gray-400">
              <p className="text-lg font-semibold">Drop components here</p>
              <p className="text-sm">
                Drag and drop from the left panel to build your form.
              </p>
            </div>
          </div>
        )}
      </div>
    </div>
  );
}
