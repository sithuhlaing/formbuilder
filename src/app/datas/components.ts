export const components = {
  input_components: {
    icon: "✏️",
    label: "Input Components",
    components: [
      {
        type: "text_input",
        icon: "📝",
        label: "Text Input",
        description: "Single line text field",
        properties: {
          label: "Text Input",
          placeholder: "Enter text",
          required: false,
        },
      },
      {
        type: "email_input",
        icon: "📧",
        label: "Email Input",
        description: "Email address field with validation",
        properties: {
          label: "Email Input",
          placeholder: "Enter email",
          required: true,
        },
      },
      {
        type: "password_input",
        icon: "🔒",
        label: "Password Input",
        description: "Password field with masking",
        properties: {
          label: "Password Input",
          placeholder: "Enter password",
          required: true,
        },
      },
      {
        type: "number_input",
        icon: "🔢",
        label: "Number Input",
        description: "Numeric input with validation",
        properties: {
          label: "Number Input",
          placeholder: "Enter number",
          required: false,
        },
      },
      {
        type: "textarea",
        icon: "📄",
        label: "Text Area",
        description: "Multi-line text input",
        properties: {
          label: "Text Area",
          placeholder: "Enter long text",
          required: false,
        },
      },
      {
        type: "rich_text",
        icon: "✍️",
        label: "Rich Text Editor",
        description: "WYSIWYG text editor",
        properties: {
          label: "Rich Text Editor",
          content: "Initial content",
        },
      },
      {
        type: "patient_id",
        icon: "🆔",
        label: "Patient ID",
        description: "Standardised patient identifier with validation",
        properties: {
          label: "Patient ID",
          placeholder: "e.g. PAT-123456",
          required: true,
          validation: [
            { type: "required", message: "Patient ID is required" },
            {
              type: "pattern",
              pattern: "^PAT-[0-9]{6}$",
              message: "Use format PAT-123456",
            },
          ],
        },
      },
      {
        type: "medical_record_number",
        icon: "📁",
        label: "Medical Record Number",
        description: "Unique MRN capture with format guardrails",
        properties: {
          label: "Medical Record Number",
          placeholder: "e.g. MRN-00000001",
          required: true,
          validation: [
            { type: "required", message: "Medical record number is required" },
            {
              type: "pattern",
              pattern: "^MRN-[0-9]{8}$",
              message: "Use format MRN-00000000",
            },
          ],
        },
      },
      {
        type: "insurance_member_number",
        icon: "💳",
        label: "Insurance Member Number",
        description: "Capture insurance membership identifiers",
        properties: {
          label: "Insurance Member Number",
          placeholder: "e.g. HMO-1234-5678",
          required: false,
          validation: [
            {
              type: "pattern",
              pattern: "^[A-Z]{2,3}-[0-9]{4}-[0-9]{4}$",
              message: "Use format HMO-1234-5678",
            },
          ],
        },
      },
      {
        type: "nhs_number",
        icon: "🏥",
        label: "NHS Number",
        description: "UK NHS compliant patient identifier",
        properties: {
          label: "NHS Number",
          placeholder: "000 000 0000",
          required: false,
          validation: [
            {
              type: "pattern",
              pattern: "^[0-9]{3}\\s?[0-9]{3}\\s?[0-9]{4}$",
              message: "Enter a 10 digit NHS number",
            },
          ],
        },
      },
    ],
  },
  selection_components: {
    icon: "☑️",
    label: "Selection Components",
    components: [
      {
        type: "select",
        icon: "🔽",
        label: "Select Dropdown",
        description: "Single selection dropdown",
        properties: {
          label: "Select Dropdown",
          options: ["Option 1", "Option 2", "Option 3"],
          required: false,
        },
      },
      {
        type: "multi_select",
        icon: "⏬",
        label: "Multi Select",
        description: "Multiple selection dropdown",
        properties: {
          label: "Multi Select",
          options: ["Option 1", "Option 2", "Option 3"],
          required: false,
        },
      },
      {
        type: "checkbox",
        icon: "✅",
        label: "Checkbox",
        description: "Single checkbox input",
        properties: {
          label: "Checkbox",
          required: false,
        },
      },
      {
        type: "radio_group",
        icon: "🔘",
        label: "Radio Group",
        description: "Single selection from multiple options",
        properties: {
          label: "Radio Group",
          options: ["Option 1", "Option 2", "Option 3"],
          required: false,
        },
      },
    ],
  },
  special_components: {
    icon: "⭐",
    label: "Special Components",
    components: [
      {
        type: "date_picker",
        icon: "📅",
        label: "Date Picker",
        description: "Calendar date selection",
        properties: {
          label: "Date Picker",
          required: false,
        },
      },
      {
        type: "file_upload",
        icon: "📤",
        label: "File Upload",
        description: "File selection and upload",
        properties: {
          label: "File Upload",
          required: false,
        },
      },
      {
        type: "signature",
        icon: "✒️",
        label: "Signature Pad",
        description: "Digital signature capture",
        properties: {
          label: "Signature Pad",
          required: false,
        },
      },
    ],
  },
  layout_components: {
    icon: "📐",
    label: "Layout Components",
    components: [
      {
        type: "horizontal_layout",
        icon: "↔️",
        label: "Horizontal Layout",
        description: "Container for side-by-side components",
        properties: {},
      },
      {
        type: "vertical_layout",
        icon: "↕️",
        label: "Vertical Layout",
        description: "Container for stacked components",
        properties: {},
      },
    ],
  },
  ui_components: {
    icon: "🎨",
    label: "UI Components",
    components: [
      {
        type: "section_divider",
        icon: "➖",
        label: "Section Divider",
        description: "Visual separator between form sections",
        properties: {},
      },
      {
        type: "button",
        icon: "🔘",
        label: "Button",
        description: "Action button for form interactions",
        properties: {
          label: "Button",
          variant: "primary",
        },
      },
      {
        type: "heading",
        icon: "🇭",
        label: "Heading",
        description: "Section heading or title",
        properties: {
          text: "Heading",
          level: "h2",
        },
      },
      {
        type: "card",
        icon: "📇",
        label: "Card Container",
        description: "Visual card container for grouping",
        properties: {},
      },
    ],
  },
};
