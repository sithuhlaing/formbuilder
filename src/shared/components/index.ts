/**
 * Shared Components - Clean Export
 */

export { Button } from './Button';
export { Modal } from './Modal';
export { ActionButton } from './ActionButton';
export { ConfirmDialog } from './ConfirmDialog';
export { DragLayer } from './DragLayer';

// Form Components
export { FormField } from './FormField';
export type { FormFieldProps } from './FormField';

// Input Components
export { 
  TextInput, 
  NumberInput, 
  Textarea, 
  Select, 
  Checkbox, 
  RadioGroup, 
  FileUpload 
} from './Input';

// Specialized Form Components
export {
  SectionDivider,
  SignatureField,
  ButtonPreview,
  HeadingPreview,
  CardPreview,
  DatePicker,
  UnknownComponent
} from './SpecializedFormComponents';

// Simple Component Renderer - Phase 4 Implementation
export { renderSimpleComponent } from '../../components/SimpleRenderer';

// Legacy components have been removed - using simplified architecture
// Current drag-drop functionality in:
// - SimpleCanvas 
// - SimpleDraggableComponent 
// - SimpleComponentPalette