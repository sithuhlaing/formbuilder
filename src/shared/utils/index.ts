
export { generateId } from './generateId';
export { classNames } from './classNames';
// debounce and deepClone removed - were unused

// Type exports for utilities
export type ComponentTypesMap = typeof COMPONENT_TYPES;

// Constants
export const COMPONENT_TYPES = {
  TEXT_INPUT: 'text_input',
  EMAIL_INPUT: 'email_input',
  PASSWORD_INPUT: 'password_input',
  NUMBER_INPUT: 'number_input',
  TEXTAREA: 'textarea',
  RICH_TEXT: 'rich_text',
  SELECT: 'select',
  MULTI_SELECT: 'multi_select',
  CHECKBOX: 'checkbox',
  RADIO_GROUP: 'radio_group',
  DATE_PICKER: 'date_picker',
  FILE_UPLOAD: 'file_upload',
  SECTION_DIVIDER: 'section_divider',
  SIGNATURE: 'signature',
  BUTTON: 'button',
  HEADING: 'heading',
  CARD: 'card',
  HORIZONTAL_LAYOUT: 'horizontal_layout',
  VERTICAL_LAYOUT: 'vertical_layout',
} as const;
