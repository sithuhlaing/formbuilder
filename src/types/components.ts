/**
 * UNIFIED COMPONENT TYPES - Phase 2 Implementation
 * Replaces: Complex interface hierarchy in core/interfaces/ComponentInterfaces.ts
 * Goal: Single, simple interface instead of 15+ intersected interfaces
 */

// Simple component types - no over-engineering
export type ComponentType = 
  // Input Components
  | 'text_input' 
  | 'email_input' 
  | 'number_input' 
  | 'textarea'
  
  // Selection Components  
  | 'select' 
  | 'radio_group' 
  | 'checkbox'
  
  // Special Components
  | 'date_picker' 
  | 'file_upload'
  
  // Layout Components
  | 'horizontal_layout' 
  | 'vertical_layout'
  
  // UI Components
  | 'button' 
  | 'heading' 
  | 'paragraph'
  | 'divider'
  | 'section_divider';

// Single, unified component interface - replacing complex hierarchy
export interface Component {
  // === Core Properties ===
  id: string;
  type: ComponentType;
  label: string;
  required?: boolean;
  
  // === Input Properties ===
  placeholder?: string;
  defaultValue?: string | number | boolean | string[];
  readOnly?: boolean;
  disabled?: boolean;
  
  // === Selection Components (select, radio, checkbox) ===
  options?: Array<{
    label: string; 
    value: string;
  }>;
  
  // === Layout Components (horizontal_layout, vertical_layout) ===
  children?: Component[];
  
  // === Simple Validation (replacing complex ValidationEngine) ===
  validation?: {
    required?: boolean;
    pattern?: string;
    minLength?: number;
    maxLength?: number;
    min?: number;
    max?: number;
    message?: string;
  };
  
  // === Simple Styling (replacing complex ComponentLayout) ===
  style?: React.CSSProperties;
  className?: string;
  
  // === Type-Specific Properties ===
  
  // For textarea
  rows?: number;
  cols?: number;
  
  // For file upload
  accept?: string;
  multiple?: boolean;
  
  // For heading
  level?: 1 | 2 | 3 | 4 | 5 | 6;
  
  // For button
  variant?: 'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info';
  size?: 'sm' | 'md' | 'lg';
  fullWidth?: boolean;
  
  // For number input
  step?: number;
  
  // For paragraph/text content
  content?: string;
  
  // === Meta Properties ===
  helpText?: string;
  fieldId?: string; // For form field identification
}

// Helper type for form field validation result
export interface ValidationResult {
  isValid: boolean;
  message?: string;
  errors?: string[];
}

// Form data structure - simple and clean
export interface FormData {
  templateName: string;
  components: Component[];
  metadata?: {
    createdAt?: string;
    updatedAt?: string;
    version?: string;
    author?: string;
  };
}

// Drop position for drag-drop operations
export interface DropPosition {
  index: number;
  parentId?: string;
  position?: 'before' | 'after' | 'inside';
}

// Drag item interface for React DnD
export interface DragItem {
  type: ComponentType;
  id?: string; // for existing components being moved
  component?: Component; // for palette items
}

// Export/Import formats
export interface FormTemplate {
  templateName: string;
  pages?: Array<{
    title: string;
    components: Component[];
  }>;
  components: Component[]; // Backward compatibility
  version: string;
}

// Default labels for component types
export const DEFAULT_COMPONENT_LABELS: Record<ComponentType, string> = {
  text_input: 'Text Input',
  email_input: 'Email Address',
  number_input: 'Number',
  textarea: 'Text Area',
  select: 'Select',
  radio_group: 'Radio Group',
  checkbox: 'Checkbox',
  date_picker: 'Date',
  file_upload: 'File Upload',
  horizontal_layout: 'Horizontal Layout',
  vertical_layout: 'Vertical Layout',
  button: 'Button',
  heading: 'Heading',
  paragraph: 'Paragraph',
  divider: 'Divider',
  section_divider: 'Section Divider'
};

// Default placeholders for input types
export const DEFAULT_PLACEHOLDERS: Partial<Record<ComponentType, string>> = {
  text_input: 'Enter text here...',
  email_input: 'Enter email address...',
  number_input: 'Enter number...',
  textarea: 'Enter text here...',
  date_picker: 'Select date...'
};

// Component type categories for UI organization
export const COMPONENT_CATEGORIES = {
  INPUT: ['text_input', 'email_input', 'number_input', 'textarea', 'date_picker', 'file_upload'] as ComponentType[],
  SELECTION: ['select', 'radio_group', 'checkbox'] as ComponentType[],
  LAYOUT: ['horizontal_layout', 'vertical_layout'] as ComponentType[],
  CONTENT: ['heading', 'paragraph', 'button', 'divider', 'section_divider'] as ComponentType[]
};

// Utility function to generate unique IDs
export function generateComponentId(prefix: string = 'comp'): string {
  return `${prefix}_${Math.random().toString(36).substr(2, 9)}`;
}

// Utility function to check if component can have children
export function canHaveChildren(componentType: ComponentType): boolean {
  return COMPONENT_CATEGORIES.LAYOUT.includes(componentType);
}

// Utility function to check if component is a form field
export function isFormField(componentType: ComponentType): boolean {
  return [...COMPONENT_CATEGORIES.INPUT, ...COMPONENT_CATEGORIES.SELECTION].includes(componentType);
}

// Utility function to get component category
export function getComponentCategory(componentType: ComponentType): string {
  for (const [category, types] of Object.entries(COMPONENT_CATEGORIES)) {
    if (types.includes(componentType)) {
      return category;
    }
  }
  return 'UNKNOWN';
}

// Type guards for better type safety
export function isInputComponent(component: Component): boolean {
  return COMPONENT_CATEGORIES.INPUT.includes(component.type);
}

export function isSelectionComponent(component: Component): boolean {
  return COMPONENT_CATEGORIES.SELECTION.includes(component.type);
}

export function isLayoutComponent(component: Component): boolean {
  return COMPONENT_CATEGORIES.LAYOUT.includes(component.type);
}

export function isContentComponent(component: Component): boolean {
  return COMPONENT_CATEGORIES.CONTENT.includes(component.type);
}