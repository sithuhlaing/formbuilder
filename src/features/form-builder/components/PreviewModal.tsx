/**
 * Preview Modal Component - Modal wrapper for form preview with multi-page support
 */

import React, { useState } from 'react';
import { Modal } from '../../../shared/components/Modal';
import { PreviewForm } from './PreviewForm';
import type { FormComponentData } from '../../../types';
import type { FormPage } from '../../../core/FormState';

interface PreviewModalProps {
  isOpen: boolean;
  onClose: () => void;
  templateName: string;
  components?: FormComponentData[];
  pages?: FormPage[];
}

export const PreviewModal: React.FC<PreviewModalProps> = ({
  isOpen,
  onClose,
  templateName,
  components = [],
  pages = []
}) => {
  const [submitModal, setSubmitModal] = useState<{
    isOpen: boolean;
    data: Record<string, any> | null;
  }>({ isOpen: false, data: null });

  const handleFormSubmit = (data: Record<string, any>) => {
    console.log('Preview form submitted:', data);
    setSubmitModal({ isOpen: true, data });
  };

  const isMultiPage = pages.length > 0;
  const totalPages = isMultiPage ? pages.length : 1;

  const modalFooter = (
    <div className="flex justify-between items-center text-sm text-gray-500">
      <span>This is a preview mode. No data will be actually submitted.</span>
      {isMultiPage && (
        <span className="font-medium">
          Multi-page form ({totalPages} page{totalPages !== 1 ? 's' : ''})
        </span>
      )}
    </div>
  );

  return (
    <Modal
      isOpen={isOpen}
      onClose={onClose}
      title={`👁️ Preview: ${templateName}`}
      size="large"
      footer={modalFooter}
    >
      <PreviewForm
        templateName={templateName}
        components={components}
        pages={pages}
        onSubmit={handleFormSubmit}
      />
      
      {/* Form Submission Success Modal */}
      <Modal
        isOpen={submitModal.isOpen}
        onClose={() => setSubmitModal({ isOpen: false, data: null })}
        title="Form Submitted Successfully"
        size="medium"
      >
        <div className="text-center py-4">
          <div className="text-green-600 text-4xl mb-4">✅</div>
          <p className="text-gray-700 mb-4">Your form has been submitted successfully!</p>
          <details className="text-left">
            <summary className="cursor-pointer text-blue-600 hover:text-blue-800 mb-2">
              View submitted data
            </summary>
            <pre className="bg-gray-100 p-3 rounded text-sm overflow-auto max-h-64">
              {JSON.stringify(submitModal.data, null, 2)}
            </pre>
          </details>
        </div>
      </Modal>
    </Modal>
  );
};