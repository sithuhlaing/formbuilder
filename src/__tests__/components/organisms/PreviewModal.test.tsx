import { vi } from 'vitest';
import React from 'react';
import { render, screen, fireEvent } from '../../utils/test-utils';
import { PreviewModal } from '../../../features/form-builder/components/PreviewModal';
import { TEST_IDS } from '../../utils/test-utils';

describe('PreviewModal', () => {
  const mockOnClose = vi.fn();
  const mockOnSubmit = vi.fn();
  
  const defaultProps = {
    isOpen: true,
    onClose: mockOnClose,
    formTitle: 'Test Form',
    pages: [{
      id: 'page-1',
      title: 'Page 1',
      components: [{
        id: 'text-1',
        type: 'text_input',
        label: 'Test Input',
        required: false,
        placeholder: 'Enter text'
      }]
    }],
    onSubmit: mockOnSubmit
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders modal when isOpen is true', () => {
    render(<PreviewModal {...defaultProps} />);
    
    expect(screen.getByRole('dialog')).toBeInTheDocument();
    expect(screen.getByText('Preview: Test Form')).toBeInTheDocument();
  });

  it('does not render when isOpen is false', () => {
    render(<PreviewModal {...defaultProps} isOpen={false} />);
    
    expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
  });

  it('calls onClose when close button is clicked', () => {
    render(<PreviewModal {...defaultProps} />);
    
    const closeButton = screen.getByRole('button', { name: /close/i });
    fireEvent.click(closeButton);
    
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('calls onClose when clicking outside the modal', () => {
    render(
      <div>
        <div data-testid="outside">Outside</div>
        <PreviewModal {...defaultProps} />
      </div>
    );
    
    fireEvent.mouseDown(screen.getByTestId('outside'));
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('renders PreviewForm with correct props', () => {
    render(<PreviewModal {...defaultProps} />);
    
    expect(screen.getByLabelText('Test Input')).toBeInTheDocument();
  });

  it('calls onSubmit when form is submitted', () => {
    render(<PreviewModal {...defaultProps} />);
    
    const submitButton = screen.getByRole('button', { name: /submit/i });
    fireEvent.click(submitButton);
    
    expect(mockOnSubmit).toHaveBeenCalledTimes(1);
  });

  it('matches snapshot when open', () => {
    const { container } = render(<PreviewModal {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
