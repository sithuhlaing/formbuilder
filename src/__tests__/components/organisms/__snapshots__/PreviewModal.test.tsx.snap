// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PreviewModal > matches snapshot when open 1`] = `
<div>
  <div
    class="modal-overlay"
  >
    <div
      class="modal max-w-4xl"
    >
      <div
        class="modal__header"
      >
        <h2>
          👁️ Preview: undefined
        </h2>
        <button
          class="modal__close"
        >
          ✕
        </button>
      </div>
      <div
        class="modal__content"
      >
        <div
          class="preview-form"
        >
          <form
            class="preview-form__content"
          >
            <div
              class="preview-field"
            >
              <label
                class="preview-field__label"
              >
                Test Input
              </label>
              <input
                class="preview-field__input "
                placeholder="Enter text"
                type="text"
                value=""
              />
            </div>
            <div
              class="preview-form__page-controls"
            >
              <div
                class="preview-form__nav-buttons"
              >
                <button
                  class="preview-form__submit"
                  type="submit"
                >
                  Submit Form
                </button>
              </div>
              <button
                class="preview-form__reset"
                type="button"
              >
                Reset Form
              </button>
            </div>
          </form>
          <div
            class="preview-form__note"
          >
            <small
              class="text-gray-500"
            >
              This is a preview mode. No data will be actually submitted.
            </small>
          </div>
        </div>
      </div>
      <div
        class="modal__footer"
      >
        <div
          class="flex justify-between items-center text-sm text-gray-500"
        >
          <span>
            This is a preview mode. No data will be actually submitted.
          </span>
          <span
            class="font-medium"
          >
            Multi-page form (
            1
             page
            )
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
